stdlib/public/SDK/Foundation/NSDictionary.swift

```diff
@@ -62,64 +62,99 @@ extension Dictionary : _ObjectiveCBridgeable {
       result = native
       return
     }
 
     if _isBridgedVerbatimToObjectiveC(Key.self) &&
        _isBridgedVerbatimToObjectiveC(Value.self) {
-      result = [Key : Value](_cocoaDictionary: d)
+      result = [Key : Value](_cocoaDictionary: x)
       return
     }
 
-    if Key.self == String.self {
+    // `Dictionary<Key, Value>` where either `Key` or `Value` is a value type
+    // may not be backed by an NSDictionary.
+    var builder = _DictionaryBuilder<Key, Value>(count: x.count)
+    var maybeDict:Dictionary? = nil
+    x.enumerateKeysAndObjects({ (anyKey: Any, anyValue: Any, stopPtr) in
+      let key = Swift._forceBridgeFromObjectiveC(
+              anyKey as AnyObject, Key.self)
+      let value = Swift._forceBridgeFromObjectiveC(
+              anyValue as AnyObject, Value.self)
       // String and NSString have different concepts of equality, so
       // string-keyed NSDictionaries may generate key collisions when bridged
       // over to Swift. See rdar://problem/35995647
-      var dict = Dictionary(minimumCapacity: d.count)
-      d.enumerateKeysAndObjects({ (anyKey: Any, anyValue: Any, _) in
-        let key = Swift._forceBridgeFromObjectiveC(
-          anyKey as AnyObject, Key.self)
-        let value = Swift._forceBridgeFromObjectiveC(
-          anyValue as AnyObject, Value.self)
+      if Key.self == String.self &&
+            !(unsafeBitCast(key, to: String.self)._isASCII) {
+         maybeDict = builder.take()
+      }
+      if maybeDict != nil {
         // FIXME: Log a warning if `dict` already had a value for `key`
-        dict[key] = value
-      })
-      result = dict
-      return
-    }
-
-    // `Dictionary<Key, Value>` where either `Key` or `Value` is a value type
-    // may not be backed by an NSDictionary.
-    var builder = _DictionaryBuilder<Key, Value>(count: d.count)
-    d.enumerateKeysAndObjects({ (anyKey: Any, anyValue: Any, _) in
-      let anyObjectKey = anyKey as AnyObject
-      let anyObjectValue = anyValue as AnyObject
-      builder.add(
-          key: Swift._forceBridgeFromObjectiveC(anyObjectKey, Key.self),
-          value: Swift._forceBridgeFromObjectiveC(anyObjectValue, Value.self))
+        maybeDict![key] = value
+      } else {
+        builder.add(
+            key: key,
+            value: value)
+      }
     })
-    result = builder.take()
+    result = maybeDict ?? builder.take()
   }
 
   public static func _conditionallyBridgeFromObjectiveC(
     _ x: NSDictionary,
     result: inout Dictionary?
   ) -> Bool {
-    let anyDict = x as [NSObject : AnyObject]
+    if let native = [Key : Value]._bridgeFromObjectiveCAdoptingNativeStorageOf(
+        x as AnyObject) {
+      result = native
+      return true
+    }
+
     if _isBridgedVerbatimToObjectiveC(Key.self) &&
        _isBridgedVerbatimToObjectiveC(Value.self) {
-      result = Swift._dictionaryDownCastConditional(anyDict)
-      return result != nil
+      result = [Key : Value](_cocoaDictionary: x)
```

Comment:
Ah, this explains the 10x speedup! Sadly, I'm afraid we'll need to get rid of it. 😭

The *ObjectiveCBridgeFromNSDictionaryAnyObject* benchmark measures `NSDictionary` → `[NSString: NSNumber]` casts, which are supposed to fail if there is even a single key or value that cannot be down-casted. 

`_dictionaryDownCastConditional` laboriously checks this, while `init(_cocoaDictionary:)` assumes the contents are guaranteed to be of the correct types already.